File: Complementar - Estruturas de dados (Didático)/Algoritmos33/Alg33.rb
=begin
♦ Algoritmo 33: Estrutura de Dados Não Linear Dinâmica (não Estática) de Árvore Binária: Árvore de Busca Binária (Binary Search Tree - BST).
=end


class No
  attr_accessor :valor, :esquerda, :direita

  def initialize(valor)
    @valor = valor
    @esquerda = nil
    @direita = nil
  end
end

class ArvoreBinaria
  def initialize
    @raiz = nil
  end

  def inserir(valor)
    @raiz = inserir_recursivo(@raiz, valor)
  end

  def inserir_recursivo(no, valor)
    if no.nil?
      return No.new(valor)
    end

    if valor < no.valor
      no.esquerda = inserir_recursivo(no.esquerda, valor)
    elsif valor > no.valor
      no.direita = inserir_recursivo(no.direita, valor)
    end

    return no
  end

  def buscar(valor)
    buscar_recursivo(@raiz, valor)
  end

  def buscar_recursivo(no, valor)
    return false if no.nil?

    if valor == no.valor
      return true
    end

    if valor < no.valor
      return buscar_recursivo(no.esquerda, valor)
    else
      return buscar_recursivo(no.direita, valor)
    end
  end

  def imprimir_em_ordem
    imprimir_em_ordem_recursivo(@raiz)
  end

  def imprimir_em_ordem_recursivo(no)
    return if no.nil?

    imprimir_em_ordem_recursivo(no.esquerda)
    print "#{no.valor} "
    imprimir_em_ordem_recursivo(no.direita)
  end
end

arvore = ArvoreBinaria.new
arvore.inserir(10)
arvore.inserir(5)
arvore.inserir(20)
arvore.inserir(3)
arvore.inserir(8)
arvore.inserir(15)

puts "Em Ordem:"
arvore.imprimir_em_ordem

puts "\n\nBuscar 8: #{arvore.buscar(8)}"
puts "Buscar 12: #{arvore.buscar(12)}"

File: Estruturas de Dados e Complexidade de Algoritmos/Algoritmos33/Alg33.php
<?php
/*
♦ Algoritmo 33: Estrutura de Dados Não Linear Dinâmica (não Estática) de Árvore Binária: Árvore de Busca Binária (Binary Search Tree - BST).
*/


class No {
    public $valor;
    public $esquerda;
    public $direita;

    public function __construct($valor) {
        $this->valor = $valor;
        $this->esquerda = $this->direita = null;
    }
}

class ArvoreBinaria {
    private $raiz;

    public function __construct() {
        $this->raiz = null;
    }

    public function inserir($valor) {
        $this->raiz = $this->inserirRecursivo($this->raiz, $valor);
    }

    private function inserirRecursivo($no, $valor) {
        if ($no === null) {
            return new No($valor);
        }

        if ($valor < $no->valor) {
            $no->esquerda = $this->inserirRecursivo($no->esquerda, $valor);
        } else if ($valor > $no->valor) {
            $no->direita = $this->inserirRecursivo($no->direita, $valor);
        }

        return $no;
    }

    public function buscar($valor) {
        return $this->buscarRecursivo($this->raiz, $valor);
    }

    private function buscarRecursivo($no, $valor) {
        if ($no === null) {
            return false;
        }

        if ($valor === $no->valor) {
            return true;
        }

        return $valor < $no->valor ? $this->buscarRecursivo($no->esquerda, $valor) : $this->buscarRecursivo($no->direita, $valor);
    }

    public function imprimirEmOrdem() {
        $this->imprimirEmOrdemRecursivo($this->raiz);
    }

    private function imprimirEmOrdemRecursivo($no) {
        if ($no !== null) {
            $this->imprimirEmOrdemRecursivo($no->esquerda);
            echo $no->valor . " ";
            $this->imprimirEmOrdemRecursivo($no->direita);
        }
    }
}

$arvore = new ArvoreBinaria();
$arvore->inserir(10);
$arvore->inserir(5);
$arvore->inserir(20);
$arvore->inserir(3);
$arvore->inserir(8);
$arvore->inserir(15);

echo "Em Ordem:\n";
$arvore->imprimirEmOrdem();
echo "\n\nBuscar 8: " . ($arvore->buscar(8) ? 'true' : 'false');
echo "\nBuscar 12: " . ($arvore->buscar(12) ? 'true' : 'false');
?>
